rfcs/2021-05-17-7469-load-balancing-k8s-aggregators.md

```diff
@@ -0,0 +1,110 @@
+
+* Vector agent to Vector aggregator
+* Datadog agent to Vector aggregator
+* Syslog agent to Vector aggregator
+
+## Out of Scope
+
+* Scaling the Kafka `source`
+* Scaling and load balancing on platforms other than Kubernetes
+* Load balancing UDP based `sources`
+
+## Motivation
+
+Today, scaling Vector horizontally (by increasing replicas) is a manual process when deployed as an aggregator. This limits Vector aggregators in both reliability and performance, causing adoption concerns for users. A single aggregator will be limited in performance by the resources that can be dedicated to it, presumably with some (currently) unknown upper bounds. Vector aims to be vendor neutral, and as such we should provide the capacity to scale and load balance across Vector aggregators regardless of environment or upstream event collectors.
+
+## Internal Proposal
+
+Include a configuration for a dedicated reverse proxy that will be deployed as part of the vector-aggregator Helm chart, as well as documented configuration and installation instructions for users that run outside of Kubernetes. We should provide basic, but functional, configurations out-of-the box to enable users to "one click" install Vector as an aggregator. The proxy should dynamically resolve downsteam Vector instances and allow users to update the balance config to provide for more consistent targets in situations that require it (aggregation transforms). I propose our initially supported proxy should be HAProxy, with the next second being NGINX or Envoy. HAProxy, compared to NGINX, provides more metrics (exposed as JSON or in Prometheus format) and has native service discovery to dynamically populate its configuration. Lua can be used with NGINX to provide service discovery, for example the [nginx-ingress-controller](https://kubernetes.github.io/ingress-nginx/).
+
+HAProxy intentionally has little support for proxying UDP, as of 2.3 there is support for forwarding syslog traffic however it doesn't allow for dynamic backend configuration greatly limiting the usability for us.
+
+Below is a basic HAProxy configuration configured to leverage service discovery in a Kubernetes cluster:
+
+```haproxy
+resolvers coredns
+    nameserver dns1 kube-dns.kube-system.svc.cluster.local:53
+    hold timeout         600s
+    hold refused         600s 
+frontend vector
+    bind *:9000
+    default_backend vector_template
+backend vector_template
+    balance roundrobin
+    option tcp-check         
+    server-template srv 10 _vector._tcp.vector-aggregator-headless.vector.svc.cluster.local resolvers coredns check
```

Comment:
Do we feel confident that we can reasonably provide a template that fits into existing environments?

Said another way: these FQDNs, etc, all depend on specific service names for the aggregators, etc.... do we feel confident our Helm chart can be tweaked well enough to support the various set-ups users may have?

Reply:
I think so, at least for the scoped sources (vector, dd, syslog) - the aggregator service will be set in the same chart so it should be easily injectable. Things get more hairy as we get out of the limited scope for defaults, but I think it can be documented enough for users to override